### DISCUSSION TITLE
Appropriate cross-validation (on study units, not time intervals) in multivariate time-series classification?

### DISCUSSION THREAD

**Discussion by jmwhyte** [alan-turing-institute/sktime#1916] [Q&A]
The sktime notebook on [multivariate time series classification](https://github.com/alan-turing-institute/sktime/blob/main/examples/03_classification_multivariate.ipynb) provides a good overview, but I'd like some further information. 

I've seen some comment on cross-validation of a (e.g. random forest) classifier by partitioning time series. In my application, this is not appropriate. However, I'd like to interrogate the behaviour of the classifier as the train-test split changes, much like someone would do in non-time-series classification. 

This command in the notebook controls the train-test split:
> X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

If I wish to evaluate the classifier for different splits, should I use a loop which uses the above and which changes 
`random_state` each time? What do you recommend? Is there some associated "K-fold validation"-type command for the purpose I describe?

**Comment by jmwhyte** [comment]
@fkiraly @lmmentel I'd appreciate some input on this. I understand that sktime takes on some of the machinery of sklearn, but it's not always apparent how to modify the former for the latter. For example, if I want to follow good practice and include hyperparameter tuning and cross-validation in my time series classification pipeline, sklearn examples such as this:
[Nested versus non-nested cross-validation](https://scikit-learn.org/stable/auto_examples/model_selection/plot_nested_cross_validation_iris.html#sphx-glr-auto-examples-model-selection-plot-nested-cross-validation-iris-py)
seem to require some specialist knowledge that isn't at all obvious. Would the developers please consider providing some sktime examples?

**Comment by fkiraly** [comment]
@jmwhyte, apologies for the delay, we have been kept busy with the 0.10.0 release.

@TonyBagnall, @MatthewMiddlehurst, @chrisholder are best to ask about time series classification. Further, the time series classification notebook has recently been reworked.

Generally, you probably want to split panels of time series (i.e., collections of time series) along *instances*, i.e., that entire time series end up in the train/test set. A hacky way would be to use `sklearn`'s `StratifiedGroupKFold` and group by time series.

`GridSearchCV` should also be compatible with time series classifiers.

@TonyBagnall et al should be able to tell you more on the current functionalities for tuning and benchmarking in time series classification.

Further, can you perhaps clarify what you are trying to do; do you want to tune, or do you want to estimate algorithmic performance? If the latter, which performance do you want to estimate: conditional (on fitting) or unconditional generalization error?

**Reply by jmwhyte** [reply]
@fkiraly thanks for taking the time to reply. I'm new to Python/ML, so I'm sorry if my lack of suitable vocabulary makes these discussions more painful than they should be. I appreciate the guidance of knowledgeable people like yourself as once I'm on the right track, I'll be able to learn the right methods.

Here's what I have put together currently - a single train/test split, followed by an attempt at five-fold cross validation:

```
steps = [
    ("concatenate", ColumnConcatenator()),
    ("classify", TimeSeriesForestClassifier(n_estimators=200, random_state=1)),
]
clf = Pipeline(steps)
clf.fit(X_train, y_train)
print("single-split classification score \n", clf.score(X_test, y_test))

print("Try 5-fold cross-validation")

start_time = time()
cv_scores = cross_val_score(clf, Xuse, y, cv=5)
elapsed_time = time() - start_time
print("elapsed time for cross-validation:", elapsed_time)
print(cv_scores)
print("%0.2f accuracy with a standard deviation of %0.2f" % (cv_scores.mean(), cv_scores.std()))
```

However, this (naive) approach neglects tuning of the classifier hyper-parameters -obviously `n_estimators` is a good place to start with this. I couldn't find particular sktime guidance, but I thought [this sklearn page](https://scikit-learn.org/stable/auto_examples/model_selection/plot_nested_cross_validation_iris.html#sphx-glr-auto-examples-model-selection-plot-nested-cross-validation-iris-py) might guide me in adding this tuning into the cross-validation process in sktime. But, experimentation showed that I didn't understand which commands were appropriate for my needs.

Your tips on sklearn's `StratifiedGroupKFold` and `GridSearchCV` might give me a start. However when I was looking at the sklearn examples, I didn't see how to adapt them for a situation where I had a `Pipeline` (as in the code snippet above), rather than just a single classifier command to feed into the cross validation. For example, a case from the quoted webpage has
`svm = SVC(kernel="rbf")`
which is later used in 
`clf = GridSearchCV)(estimator=svm, param_grid=p_grid, cv=outer_cv)`
and I couldn't see how to adapt this for the `Pipeline` situation. This confusion is why I was hopeful of some simple examples to demonstrate the key principles. @TonyBagnall, @MatthewMiddlehurst, @chrisholder I would appreciate any direction you can provide on these matters.

**Reply by jmwhyte** [reply]
@fkiraly Any thought on when someone can look at this matter?

I've worked through some docs, and eliminated some errors. I might be missing something when you wrote:
"A hacky way would be to use sklearn's StratifiedGroupKFold and group by time series."
I've had a look at the cv.split() loop [here](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.StratifiedGroupKFold.html) and tried to smoosh something together. 

Suppose we begin with the example "Time Series Classification" from [here](https://www.sktime.org/en/stable/get_started.html)

    from sktime.classification.kernel_based import TimeSeriesForestClassifier
    % NOTE: this line in the example was incorrect 
    % from sktime.classification.kernel_based import TimeSeriesForestClassifier

    from sktime.datasets import load_arrow_head
    from sklearn.model_selection import train_test_split
    from sklearn.metrics import accuracy_score
    X, y = load_arrow_head()
    X_train, X_test, y_train, y_test = train_test_split(X, y)
    classifier = TimeSeriesForestClassifier()
    classifier.fit(X_train, y_train)
    y_pred = classifier.predict(X_test)
    accuracy_score(y_test, y_pred)

output presented 0.8679245283018868

What do experts think of the following attempt to combine hyper-parameter tuning and cross-validation with a stratified approach to making the train/test split?  (This is adapted from my recent work, so possibly some irrelevances)
EDIT: StratifiedGroupKFold used now, results are a little better than before.

    import numpy as np
    from sklearn.model_selection import train_test_split
    from sklearn.pipeline import Pipeline

    from sktime.classification.interval_based import TimeSeriesForestClassifier
    from sktime.transformations.panel.compose import ColumnConcatenator

    from sklearn.model_selection import StratifiedGroupKFold
    from time import time

    nest_grid = range(160,220,5)
    NUM_TRIALS = len(nest_grid)
    val_fold = 5

    results_length = NUM_TRIALS * val_fold

    groups = range(0,len(y))
    % Array to store scores
    predicted_score = np.zeros(val_fold)

    cv = StratifiedGroupKFold(n_splits=val_fold)

    start_time = time()

    % Loop for each trial (a particular parameter value), for which five-fold cross-validation occurs
    for i in range(NUM_TRIALS):

    steps = [
    ("concatenate", ColumnConcatenator()),
    ("classify", TimeSeriesForestClassifier(n_estimators=nest_grid[i]))]
    clf = Pipeline(steps)

    
    j=0
    for train_idxs, test_idxs in cv.split(X, y, groups):
        X_train, X_test = X.iloc[train_idxs][X.columns], X.iloc[test_idxs][X.columns]
        y_train, y_test = y[train_idxs], y[test_idxs]

        clf.fit(X_train, y_train)
        predicted_score[j] = clf.score(X_test, y_test)
        j = j+1

    print("For n_estimators=", nest_grid[i], "All scores", predicted_score)
    print("Av scores of {:6f} with std. dev. of {:6f}.".format(
    predicted_score.mean(), predicted_score.std())
    )
    print("")

    print("")
    elapsed_time = time() - start_time
    print("elapsed time for cross-validation/hyperparameter grid search:", elapsed_time)

Output:

For n_estimators= 160 All scores [0.80952381 0.80952381 0.86046512 0.88095238 0.92857143]
Av scores of 0.857807 with std. dev. of 0.045195.

For n_estimators= 165 All scores [0.78571429 0.83333333 0.88372093 0.88095238 0.9047619 ]
Av scores of 0.857697 with std. dev. of 0.042914.

For n_estimators= 170 All scores [0.76190476 0.85714286 0.86046512 0.85714286 0.9047619 ]
Av scores of 0.848283 with std. dev. of 0.046811.

For n_estimators= 175 All scores [0.76190476 0.80952381 0.86046512 0.9047619  0.92857143]
Av scores of 0.853045 with std. dev. of 0.061048.

For n_estimators= 180 All scores [0.76190476 0.83333333 0.88372093 0.88095238 0.92857143]
Av scores of 0.857697 with std. dev. of 0.056588.

For n_estimators= 185 All scores [0.76190476 0.83333333 0.86046512 0.88095238 0.9047619 ]
Av scores of 0.848283 with std. dev. of 0.049174.

For n_estimators= 190 All scores [0.78571429 0.83333333 0.8372093  0.9047619  0.92857143]
Av scores of 0.857918 with std. dev. of 0.051832.

For n_estimators= 195 All scores [0.76190476 0.78571429 0.86046512 0.9047619  0.9047619 ]
Av scores of 0.843522 with std. dev. of 0.059650.

For n_estimators= 200 All scores [0.78571429 0.85714286 0.86046512 0.88095238 0.9047619 ]
Av scores of 0.857807 with std. dev. of 0.039863.

For n_estimators= 205 All scores [0.78571429 0.83333333 0.86046512 0.85714286 0.9047619 ]
Av scores of 0.848283 with std. dev. of 0.038872.

For n_estimators= 210 All scores [0.76190476 0.83333333 0.86046512 0.85714286 0.9047619 ]
Av scores of 0.843522 with std. dev. of 0.046879.

For n_estimators= 215 All scores [0.76190476 0.83333333 0.86046512 0.88095238 0.9047619 ]
Av scores of 0.848283 with std. dev. of 0.049174.


elapsed time for cross-validation/hyperparameter grid search: 143.7835509777069

**Reply by fkiraly** [reply]
> @fkiraly Any thought on when someone can look at this matter?

Apologies for the wait, I was hoping someone more closely involved with the time series classification module would find time to respond.

I've looked at the matter a bit now and I think:
* I don't think there is a good way to do grid search for the TSC module, like there is for forecasting or `sklearn`. It seems the TSC module authors have primarily focused on individual algorithms, not on composition. Please correct me if I'm wrong, @TonyBagnall et al.
* your approach seems sound, for tuning - as long as you apply them for prediction on unseen data, i.e., rows that are not in `X` and `y`. (usual caveats on train/validate/test apply for time series considered as iid samples) The code block you wrote should be the `_fit` function of the tuning algorithm.

Given this, I think we should have sth like this.
Would you be willing to add your approach as grid search for the TSC module?
(assuming I'm right and this does not exist yet, which I strongly suspect after looking into it)

**Reply by fkiraly** [reply]
I opened an issue based on the assumptions above:
https://github.com/alan-turing-institute/sktime/issues/2116

**Reply by jmwhyte** [reply]
@fkiraly I'm very happy to contribute (and be a part of discussions) to TSC module if others see value in it, and EXTREMELY happy to take advice. (Unfortunately I've taken on Python/sklearn/sktime with a basic ML understanding - I have only read about train/test, not train/validate/test -  so am very happy to have pointers as I don't always know how things should work. I tried your earlier suggestion of using `StratifiedGroupKFold` originally, but found that I could have some groups which were all 0-label entries. So, I changed to `StratifiedKFold`, which seemed to give me a diversity of elements in the train/test split.)

**Reply by fkiraly** [reply]
Ah, great! Re. train-validate-test, it's a very important point when it comes to tuning, since you can run into overfitting or "evaluating on the test set" issues if not aware! See here https://en.wikipedia.org/wiki/Training,_validation,_and_test_sets, or section 7 of https://hastie.su.domains/ElemStatLearn/printings/ESLII_print12_toc.pdf for a more technical exposition.

>I tried your earlier suggestion of using StratifiedGroupKFold originally, but found that I could have some groups which were all 0-label entries. So, I changed to StratifiedKFold, which seemed to give me a diversity of elements in the train/test split.)

I think this is not exactly right, what you are doing - you want to stratify by the *instance index* of your time series, rather than by label.
I missed that when glossing over your code. You want entire series to end up either in the test, validation, or training set.

**Reply by jmwhyte** [reply]
Thanks for the references @fkiraly! Also, I have followed your suggestion with StratifiedGroupKFold/groups and updated the code snippet above. It seems to work as expected now.

**Reply by fkiraly** [reply]
Great! If you want to contribute this, it should be wrapped in the [classifier extension template](https://github.com/alan-turing-institute/sktime/blob/main/extension_templates/classification.py) which has "todo"s for every place you need to add sth or change code.
This would be a composite so might be slightly fiddly, but I hope the instructions are good enough :-)

For generic dev infos and how to get involved, see here: https://www.sktime.org/en/stable/developer_guide.html#developer-guide

**Reply by jmwhyte** [reply]
I will get on to this when I get a chance.
I think I need to learn a bit more, like how to set a random seed for reproducibility.
Maybe there are other subtleties .... (I didn't quite understand the [@MatthewMiddlehurst](https://github.com/MatthewMiddlehurst) comment on using `cross_val_predict` given the concerns about groups/stratification.)
(I'm really looking forward to the new, improved tutorial!)

**Reply by jmwhyte** [reply]
@fkiraly I spent some time on your classifier template, but didn't really understand the process (or the errors returned):
---> MyTimeSeriesClassifier.fit(X,y)

TypeError: fit() missing 1 required positional argument: 'y'

I would say I need to look at an example to see how matters (parameters?) are handled as I'm missing too much background currently. However, in [#2116](https://github.com/alan-turing-institute/sktime/issues/2116) it seems as though this issue has become more of a priority, so maybe I should wait?

**Reply by fkiraly** [reply]
hm, happy to help with what you are trying there.
The template is meant to be filled in, then you have to instantiate the classifier before running `fit`etc:

```python
my_classifier = MyTimeSeriesClassifier(param1, param2, etc)
my_classifier.fit(X, y)
y_pred = my_classifier.predict(X_new)
```

if you have done everything right in filling in the template, this should work.

Regarding examples: every classifier currently implemented in `sktime` is an example for what to fill in and how, look at `TSFreshClassifier` (very simple) or `TemporalDictionaryEnsemble` (more complex)

**Reply by jmwhyte** [reply]
@fkiraly thanks for the help - you spotted my omission. I now have output from the template, and will go through and add further details. 
In my original loop, I passed the mean and standard deviation of cross-validation scores to matplot to show an errorbar plot to guide interpretation. As a design query: would you suggest that I make such a plot in the new MyTimeSeriesClassifier where the results are produced? Or, would you say I should return the mean/st_dev arrays and process them outside of the classifier call? (If this option, I'll need to learn how to do it, as the usual "return" from a function doesn't seem appropriate here.) 
I will certainly look at the examples you listed so I can gain greater familiarity with the system. Thanks again.
